<font color=blue> Привет еще раз. Спасибо, что доделал работу. Мои комментарии на этот раз выделены синим цветом.
    
---


# Исследование надёжности заёмщиков

## Описание проекта
Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.



### Описание данных
- children — количество детей в семье
- days_employed — общий трудовой стаж в днях
- dob_years — возраст клиента в годах
- education — уровень образования клиента
- education_id — идентификатор уровня образования
- family_status — семейное положение
- family_status_id — идентификатор семейного положения
- gender — пол клиента
- income_type — тип занятости
- debt — имел ли задолженность по возврату кредитов
- total_income — ежемесячный доход
- purpose — цель получения кредита

### Краткий план настоящей работы

- определите и заполните пропущенные значения;
- исследовать и заменить при необходимости типы данных;
- исследовать и удалить дубликаты;
- выделите леммы с целями получения кредита;
- провести категоризацию данных;
- сделать вывод по вопросу: Есть ли зависимость между наличием детей и возвратом кредита в срок?;
- сделать вывод по вопросу: Есть ли зависимость между семейным положением и возвратом кредита в срок?;
- сделать вывод по вопросу: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?;
- сделать вывод по вопросу: Как разные цели кредита влияют на его возврат в срок?;
- сделать общий вывод по проделанной работе.

---
## Комментарий от наставника

Привет! Спасибо, что прислал задание:) В целом, работа производит неплохое впечатление, однако обрати внимание на некоторые мои комментарии. Они обозначены пометкой **Комментарий от наставника**. Пострайся учесть эти комментарии.

---


---
## Комментарий от наставника


Правильно, что есть краткое вступление в работу, описание того, что надо делать. В работе необходимо приводить краткий план того, что надо сделать(если этот план имеется), а также информацию о входных данных: какие столбцы есть в таблице, их названия и какую информацию они несут (см. пример). Также название работы лучше отображать в отдельной ячейке и делать крупный шрифт (заголовок). Так работа выглядит презентабельно.

<font color=red> Пример: </font>
---
Описание данных: 
- children — количество детей в семье
- days_employed — общий трудовой стаж в днях
- dob_years — возраст клиента в годах
- education — уровень образования клиента
- education_id — идентификатор уровня образования
- ...
    
    
---

### Шаг 1. Откройте файл с данными и изучите общую информацию. 

In [1]:
#импорт библиотек используемых в проекте
import pandas as pd

from pymystem3 import Mystem
from collections import Counter

import numpy as np

In [2]:
data = pd.read_csv('/datasets/data.csv')

#посмотрим размер набора данных
print('')
print('--------------------------------------------------')
print('Размер набора данных, количество строк и столбцов соответственно')
print('--------------------------------------------------')
print(data.shape)
print('--------------------------------------------------')


--------------------------------------------------
Размер набора данных, количество строк и столбцов соответственно
--------------------------------------------------
(21525, 12)
--------------------------------------------------


In [3]:
#посмотрим наименования столбцов
print('')
print('--------------------------------------------------')
print('Наименование столбцов набора данных')
print('--------------------------------------------------')
data.columns


--------------------------------------------------
Наименование столбцов набора данных
--------------------------------------------------


Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

In [4]:
#выводим информацию о наборе данных
print('')
print('--------------------------------------------------')
print('Краткая информация о наборе данных')
print('--------------------------------------------------')
data.info()


--------------------------------------------------
Краткая информация о наборе данных
--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [5]:
#найдем количество грубых дубликатов
print('')
print('--------------------------------------------------')
print('Количество грубых дубликатов в наборе данных:', data.duplicated().sum())
print('--------------------------------------------------')


--------------------------------------------------
Количество грубых дубликатов в наборе данных: 54
--------------------------------------------------


In [6]:
#выведем на экран случайные 10 записей набора данных
data_sample_10 = data.sample(10)
data_sample_10

children  days_employed  dob_years education  education_id  \
3669          0  371402.500268         64   среднее             1   
1363         -1   -1195.264956         55   СРЕДНЕЕ             1   
5036          0            NaN         30    высшее             0   
13142         0            NaN         61   среднее             1   
9520          1    -275.397370         33   среднее             1   
3467          1   -1262.803022         40   среднее             1   
4753          0   -7288.363624         47    высшее             0   
9512          0   -1769.095530         25   среднее             1   
19005         1    -937.168917         28   среднее             1   
7894          2            NaN         44   среднее             1   

          family_status  family_status_id gender income_type  debt  \
3669     вдовец / вдова                 2      F   пенсионер     0   
1363    женат / замужем                 0      F   компаньон     0   
5036   гражданский брак                 1      F   компаньон     0   
13142         в разводе                 3      F   пенсионер     0   
9520    женат / замужем                 0      F   компаньон     0   
3467    женат / замужем                 0      M   сотрудник     0   
4753    женат / замужем                 0      F   компаньон     0   
9512   гражданский брак                 1      M   сотрудник     0   
19005   женат / замужем                 0      F   сотрудник     0   
7894    женат / замужем                 0      F   сотрудник     0   

        total_income                             purpose  
3669   132826.803626                        недвижимость  
1363    69550.699692              профильное образование  
5036             NaN          строительство недвижимости  
13142            NaN                сделка с автомобилем  
9520   287158.558110             покупка жилья для сдачи  
3467   485924.953268                   операции с жильем  
4753   182572.858051                               жилье  
9512   202884.713165                покупка недвижимости  
19005   93291.685781                сделка с автомобилем  
7894             NaN  на покупку подержанного автомобиля

### Вывод

В наборе данных 21525 строк и 12 стобцов, наименование столбцов однородны по регистру. 
В столбцах 'days_employed', 'total_income', 'children' присутствуют отрицательные значения.
В столбце 'education' данные представлены в разном регистре.
В столбце 'days_employed', 'total_income' присутствуют пропуски данных.
Грубых дубликатов - 54 шт., после обработки пропусков и приведения содержимого столбцов к единому регистру это число может измениться.

---
## Комментарий от наставника

У меня имеется пара комментариев по данному шагу:
- желательно выводить около 10 строчек таблицы. Меньше не рекомендуется, можно не увидеть структуру данных;
- считывание данных и импорт необходимых библиотек лучше проводить в разных ячейках;
- таблицы на печать выводи без использования метода print();
- следует давать переменным осознанные имена. Таблица про клиентов - название надо бы выбрать соответствующее;
- Разбивай код на большее число ячеек;
- Подумай также о возможных причинах появления пропусков, а также о том, являются ли они случайными или нет.

Первый взгляд на таблицу выполнен.

---

### Шаг 2. Предобработка данных

### Обработка пропусков

In [7]:
#избавимся от отрицательных значений
def gone_minus(i):
    if i < 0:
        return i*-1
    return i

data['days_employed'] = data['days_employed'].apply(gone_minus)
data['total_income'] = data['total_income'].apply(gone_minus)
data['children'] = data['children'].apply(gone_minus)

##### Выясняем причину пропусков

In [8]:
#для выяснения причин создадим новую ветку анализа, скопируем набор данных в новую переменную data_check_nan
data_check_nan = data.copy()

#для порядка, переведем содержимое основных столбцов к единому виду, в нижний регистр
data_check_nan['education'] = data_check_nan['education'].str.lower()
data_check_nan['family_status'] = data_check_nan['family_status'].str.lower()
data_check_nan['income_type'] = data_check_nan['income_type'].str.lower()
data_check_nan['purpose'] = data_check_nan['purpose'].str.lower()

#заполним пропуски в столбцах значением 0
data_check_nan['total_income'] = data_check_nan['total_income'].fillna(value=0)
data_check_nan['days_employed'] = data_check_nan['days_employed'].fillna(value=0)

#сменим тип данных
data_check_nan['days_employed'] = data_check_nan['days_employed'].astype('int')
data_check_nan['total_income'] = data_check_nan['total_income'].astype('int')

#уберем дубликаты
data_check_nan = data_check_nan.drop_duplicates().reset_index(drop=True)

#делаем срез данных с значением 0 в столбце total_income и days_employed и сравниваем их
data_0_total_income = data_check_nan.query('total_income == 0')
data_0_days_employed = data_check_nan.query('days_employed == 0')
difference_locations = np.where(data_0_total_income != data_0_days_employed)
#difference_locations
#данные в срезах идентичны - это значит что если есть пропуск по столбцу total_income, 
# то соответственно есть и по столбцу days_employed в этой же строке

#соберем сводную таблицу по уникальным значениям income_type, столбцы соберем по debt
data_pivot_0 = data_0_total_income.pivot_table(index=['income_type'], columns='debt', aggfunc={'debt':'count'})
#data_pivot_0
#порядок значений характерен для исходного набора данных

#сравним количество значений по income_type и family_status 
#для пропущенных значений и для всего набора данных
#print(data_0_total_income['income_type'].value_counts())
#print('--------------------------------------------------')
#print(data['income_type'].value_counts())
#print('--------------------------------------------------')
#print(data_0_total_income['family_status'].value_counts())
#print('--------------------------------------------------')
#print(data['family_status'].value_counts())

вывод: набор данных с пропущенными значениями характерен всему исходному набору пропуски случайны, причина их возникономения, вероятнее всего в не корректном вводе либо хранении данных

In [9]:
#заполняем пропуски

#этот код я нашел на GitHub, 
#хотел найти решение которое дало бы мне универсальный инструмент заполнения пропусков
#значениями медианы или среднего по условию причастности к определенной группе
name_for_isnull = 'total_income'
data.loc[data[name_for_isnull].isnull(), name_for_isnull] = data.groupby('income_type')[name_for_isnull].transform('median')
name_for_isnull_1 = 'days_employed'
data.loc[data[name_for_isnull_1].isnull(), name_for_isnull_1] = data.groupby('income_type')[name_for_isnull_1].transform('median')


In [10]:
#проверим на наличие пропусков в базе
print('--------------------------------------------------')
print('Проверка на наличие пропусков данных в базе:')
print('по столбцу days_employed (трудовой стаж в днях)')
print('--------------------------------------------------')
print(data[data['days_employed'].isnull()].head())
print('--------------------------------------------------')
print('Проверка на наличие пропусков данных в базе:')
print('по столбцу total_income (доход в месяц)')
print('--------------------------------------------------')
print(data[data['total_income'].isnull()].head())
print('--------------------------------------------------')

--------------------------------------------------
Проверка на наличие пропусков данных в базе:
по столбцу days_employed (трудовой стаж в днях)
--------------------------------------------------
Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []
--------------------------------------------------
Проверка на наличие пропусков данных в базе:
по столбцу total_income (доход в месяц)
--------------------------------------------------
Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []
--------------------------------------------------


In [11]:
#найдем количество грубых дубликатов
print('')
print('--------------------------------------------------')
print('Количество грубых дубликатов в наборе данных:', data.duplicated().sum())
print('--------------------------------------------------')


--------------------------------------------------
Количество грубых дубликатов в наборе данных: 54
--------------------------------------------------


### Вывод

Пропуски в базе данных были выявлены в столбцах 'days_employed' (трудовой стаж в днях) и 'total_income'(доход в месяц).
Вероятнее всего пропуски образовались из за не корректного ввода данных физическим лицом при заполнении анкеты, либо данные не корректно хранятся и обновляются.
Для заполнения пропусков было принято решение использовать медианные значения, вычисленные для каждой группы по столбцу income_type. 

<font color=blue> Пропуски обработаны верно.
    
---


---
## Комментарий от наставника

Лучше использовать заполнение пропусков в соответствии с типом дохода. Тогда пенсионеры будут получать также, как и другие пенсионеры и т.д. Все-таки в данных есть достаточное число групп, заполнять пропуски в стаже, скажем, у студентов и пенсионеров одним и тем же значением не очень то правильно. То же самое касается и столбца с доходами: пропуски у предпринимателя и пенсионера заполняются одним числом, что вызывает вопросы. Подумай, как лучше обработать пропуски.

---

### Замена типа данных

In [12]:
#изменяем тип данных
data['days_employed'] = data['days_employed'].astype('int')
data['total_income'] = data['total_income'].astype('int')
data['dob_years'] = data['dob_years'].astype('int')

#data['total_income'] = pd.to_numeric(data['total_income'], errors='raise')

print(data.info())

#найдем количество грубых дубликатов
print('')
print('--------------------------------------------------')
print('Количество грубых дубликатов в наборе данных:', data.duplicated().sum())
print('--------------------------------------------------')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB
None

--------------------------------------------------
Количество грубых дубликатов в наборе данных: 54
--------------------------------------------------


### Вывод

Для лучшей наглядности и удобства работы с данными и общей логики изменим вещественные типы данных на целочисленные, ведь количество дней лучше смотреть когда они целые. 
Изменения проводим методом astype('int').

Не совсем понял зачем здесь использовать метод to_numeric(), у меня в наборе данных нет нужды строковые значения переводить в числовые.

---
## Комментарий от наставника

Исходя из общей информации о таблице мы видим, что два столбца имеют вещественный тип данных, который и надо заменить на целочисленный. Это замена произведена верно. Также хорошо бы освоить метод to_numeric(). 

---

<font color=blue> Данный метод и не надо использовать. Но он пригодится в дальнейших проектах.
    
---


### Обработка дубликатов

In [13]:
#приведем к нижнему регистру
data['education'] = data['education'].str.lower()
data['family_status'] = data['family_status'].str.lower()
data['income_type'] = data['income_type'].str.lower()
data['purpose'] = data['purpose'].str.lower()
print('--------------------------------------------------')

#очистим набор данных от грубых дубликатов с новой индексацией
data = data.drop_duplicates().reset_index(drop=True)

print('')
print('--------------------------------------------------')
print('Информация о базе после обработки')
print('--------------------------------------------------')
print(data.info())
print('--------------------------------------------------')

#найдем количество грубых дупликатов
print('')
print('--------------------------------------------------')
print('Количество грубых дубликатов в наборе данных:', data.duplicated().sum())
print('--------------------------------------------------')

--------------------------------------------------

--------------------------------------------------
Информация о базе после обработки
--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 12 columns):
children            21454 non-null int64
days_employed       21454 non-null int64
dob_years           21454 non-null int64
education           21454 non-null object
education_id        21454 non-null int64
family_status       21454 non-null object
family_status_id    21454 non-null int64
gender              21454 non-null object
income_type         21454 non-null object
debt                21454 non-null int64
total_income        21454 non-null int64
purpose             21454 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB
None
--------------------------------------------------

--------------------------------------------------
Количество грубых дубликатов в наборе данных: 0


### Вывод

Продолжаем работать над материалом для исследования. Избавляемся от дубликатов методом drop_duplicates, но предварительно обработали содержимое текстовых столбцов - привели их к единому регистру, количество дубликатов подскочило с 54 до 71, но вскоре упало до 0 (drop_duplicates появился на площадке).

---
## Комментарий от наставника

Дубликаты вызывают смещение финальных результатов, поэтому их и надо удалять. Конечно, в данных нет уникального идентификатора, однако надо уметь удалять дубликаты. Метод drop_duplicates() использован верно.

---

### Лемматизация

In [14]:
m = Mystem()

print('')
print('Список лемм из столбца purpose (цель получения кредита)')
print('--------------------------------------------------')

#создаем список с леммами
lemmas = []
for index in data['purpose']:
    lemmas += m.lemmatize(index)
Counter(lemmas)


Список лемм из столбца purpose (цель получения кредита)
--------------------------------------------------


Counter({'покупка': 5897,
         ' ': 33570,
         'жилье': 4460,
         '\n': 21454,
         'приобретение': 461,
         'автомобиль': 4306,
         'дополнительный': 906,
         'образование': 4013,
         'сыграть': 765,
         'свадьба': 2324,
         'операция': 2604,
         'с': 2918,
         'на': 2222,
         'проведение': 768,
         'для': 1289,
         'семья': 638,
         'недвижимость': 6351,
         'коммерческий': 1311,
         'жилой': 1230,
         'строительство': 1878,
         'собственный': 635,
         'подержать': 478,
         'свой': 2230,
         'со': 627,
         'заниматься': 904,
         'сделка': 941,
         'подержанный': 486,
         'получение': 1314,
         'высокий': 1374,
         'профильный': 436,
         'сдача': 651,
         'ремонт': 607})

In [15]:
#создаем функцию которая присваивает категорию цели кредита каждой строке набора данных на основании содержимого столбца 'purpose'
# полученную категорию добавляем в новый столбец 'purpose_class'
def lemma_purpose(row):
    lemma_purposes = m.lemmatize(row['purpose'])
    if 'недвижимость' in lemma_purposes:
        return 'недвижимость'
    if 'свадьба' in lemma_purposes:
        return 'свадьба'
    if 'автомобиль' in lemma_purposes:
        return 'автомобиль'
    if 'жилье' in lemma_purposes:
        return 'недвижимость'
    if 'образование' in lemma_purposes:
        return 'образование'
    return 'что_то_другое'

data['purpose_class'] = data.apply(lemma_purpose, axis=1)

print('')
print('Сколько раз встречается выделенный класс в наборе данных')
print('--------------------------------------------------')
#посчитаем количество каждой категории в наборе данных
print(data['purpose_class'].value_counts())
print('')
print('--------------------------------------------------')


Сколько раз встречается выделенный класс в наборе данных
--------------------------------------------------
недвижимость    10811
автомобиль       4306
образование      4013
свадьба          2324
Name: purpose_class, dtype: int64

--------------------------------------------------


### Вывод

Для проведение дальнейшего анализа нам необходимо привести к какому-то стандарту данные в столбце 'purpose', для этого выделим из него леммы, выделим группы и назначим каждой строке эту самую группу.

Леммы сгруппированы по основным крупным потребительским группам, выбранным на основе результатов работы Counter(lemmas), это - недвижимость, автомобили, образование и свадьба.

Посчитали сколько раз в новом столбце 'purpose_class' встречается каждая группа. Оказалось что практически половина строк имеет цель кредита недвижимость. Анализируем дальше. 

---
## Комментарий от наставника

Все верно, метод lemmatize() использован правильно. Хорошо бы на данном этапе подробно объяснить выбор лемм.

---

### Категоризация данных

#### Присвоим категорию по возрасту клиента

In [16]:
#присвоим каждому клиенту возрастную группу
def age_group(age):
        if age < 18:
                return 'дети'
        if age <= 64:
                return 'взрослые'
        return 'пенсионеры'
    
data['age_group'] = data['dob_years'].apply(age_group)
#data.sample(10)

#посчитаем представителей каждой группы
data['age_group'].value_counts()

взрослые      20458
пенсионеры      895
дети            101
Name: age_group, dtype: int64

Обнаружили группу детей, что не возможно, так как дети не могу взять кредит в банке.
Сделаем срез данных по усовию - возраст меньше 18.

In [17]:
data.query('dob_years < 18')

children  days_employed  dob_years education  education_id  \
99            0         346541          0   среднее             1   
149           0           2664          0   среднее             1   
270           3           1872          0   среднее             1   
578           0         397856          0   среднее             1   
1040          0           1158          0    высшее             0   
...         ...            ...        ...       ...           ...   
19770         0           1574          0   среднее             1   
20397         0         338734          0   среднее             1   
20512         0         331741          0   среднее             1   
21110         2            108          0    высшее             0   
21243         0           1268          0   среднее             1   

               family_status  family_status_id gender income_type  debt  \
99           женат / замужем                 0      F   пенсионер     0   
149                в разводе                 3      F   сотрудник     0   
270          женат / замужем                 0      F   сотрудник     0   
578          женат / замужем                 0      F   пенсионер     0   
1040               в разводе                 3      F   компаньон     0   
...                      ...               ...    ...         ...   ...   
19770        женат / замужем                 0      F   сотрудник     0   
20397        женат / замужем                 0      F   пенсионер     0   
20512  не женат / не замужем                 4      F   пенсионер     0   
21110        женат / замужем                 0      M   компаньон     0   
21243  не женат / не замужем                 4      M   сотрудник     0   

       total_income                                 purpose purpose_class  \
99            71291                              автомобиль    автомобиль   
149           70176                       операции с жильем  недвижимость   
270          102166                            ремонт жилью  недвижимость   
578           97620  строительство собственной недвижимости  недвижимость   
1040         303994                         свой автомобиль    автомобиль   
...             ...                                     ...           ...   
19770        142594                                   жилье  недвижимость   
20397        259193                    покупка своего жилья  недвижимость   
20512        129788                            недвижимость  недвижимость   
21110        240702        строительство жилой недвижимости  недвижимость   
21243        152419                       операции с жильем  недвижимость   

      age_group  
99         дети  
149        дети  
270        дети  
578        дети  
1040       дети  
...         ...  
19770      дети  
20397      дети  
20512      дети  
21110      дети  
21243      дети  

[101 rows x 14 columns]

Из среза данных видно что возраст всей группы "дети" это ошибка, он равен 0, пренебрежем этой группой (её численность составляет 0,4 % от всего набора данных), присвоив по значени 0 в столбце 'dob_years' группу 'взрослые'. 

In [18]:
#присвоим каждому клиенту возрастную группу
def age_group(age):
        if age < 0:
                return 'дети'
        if age <= 64:
                return 'взрослые'
        return 'пенсионеры'
    
data['age_group'] = data['dob_years'].apply(age_group)
#data.sample(10)

#посчитаем представителей каждой группы
data['age_group'].value_counts()

взрослые      20559
пенсионеры      895
Name: age_group, dtype: int64

#### Присвоим категорию по уровню дохода

In [19]:
#посмотрим разброс значений по столбцу total_income (доходы в месяц)
print('')
print('--------------------------------------------------')
print('Информация по столбцу total_income (доходы в месяц)')
print('--------------------------------------------------')
print(data['total_income'].describe())
print('')
print('--------------------------------------------------')


--------------------------------------------------
Информация по столбцу total_income (доходы в месяц)
--------------------------------------------------
count    2.145400e+04
mean     1.653196e+05
std      9.818730e+04
min      2.066700e+04
25%      1.076230e+05
50%      1.425940e+05
75%      1.958202e+05
max      2.265604e+06
Name: total_income, dtype: float64

--------------------------------------------------


In [20]:
#на основании полученной информации напишем функцию для категоризации уровня доходов на группы
def income_level(i):
    if i < 50000:
        return '1_до 50 т. в мес.'
    if i > 50000 and i < 80000:
        return '2_от 50 до 80 т. в мес.'
    if i > 80000 and i < 150000:
        return '3_от 80 до 150 т. в мес.'
    if i > 150000 and i < 250000:
        return '4_от 150 до 250 т. в мес.'
    if i > 250000 and i < 350000:
        return '5_от 250 до 350 т. в мес.' 
    if i > 350000 and i < 450000:
        return '6_от 350 до 450 т. в мес.'
    if i > 450000 and i < 550000:
        return '7_от 450 до 550 т. в мес.'
    if i > 550000 and i < 650000:
        return '8_от 550 до 650 т. в мес.' 
    if i > 650000 and i < 750000:
        return '9_от 650 до 750 т. в мес.'
    if i > 750000 and i < 850000:
        return '10_от 750 до 850 т. в мес.'
    if i > 850000 and i < 950000:
        return '11_от 850 до 950 т. в мес.' 
    if i > 1000000:                
        return '12_ выше 1 млн. в мес'
#добавим столбец в набор данных и присвоем каждой строке группу дохода
data['income_level'] = data['total_income'].apply(income_level)

#посчитаем представителей каждой группы
data['income_level'].value_counts()

3_от 80 до 150 т. в мес.      9347
4_от 150 до 250 т. в мес.     7018
5_от 250 до 350 т. в мес.     1954
2_от 50 до 80 т. в мес.       1904
6_от 350 до 450 т. в мес.      526
1_до 50 т. в мес.              372
7_от 450 до 550 т. в мес.      179
8_от 550 до 650 т. в мес.       66
9_от 650 до 750 т. в мес.       39
12_ выше 1 млн. в мес           25
10_от 750 до 850 т. в мес.      12
11_от 850 до 950 т. в мес.       9
Name: income_level, dtype: int64

### Вывод

В набор данных добавлено 2 столбца с категориями, один столбец в котором присвоена возрастная категория, второй - категория по уровню дохода. 

На основе полученной категоризации по возрасту, можно сказать что в наборе данных пенсионеры составляют 4.18 %, весь остальной набор это взрослые люди в возрасте от 18 дло 64 лет.

Категоризация по уровню дохода. Так как понятия средний, высший и нижший класс, которыми люди привыкли оперировать в повседневной жизни довольно размытые, было принято решение просмотреть информацию о значениях в столбце 'total_income' путем применения метода describe(). На основании полученных значений о максимуме, минимуме и среднем, было принято решение "нарезать" уровень дохода с ориентировочным шагом в 100 тыс. руб.

Так же на этапе выбора лемм добавлен столбец 'purpose class' с категориями целей кредита.

---
## Комментарий от наставника

Исходя из чего выбран такой способ категоризации? Стоит подробно описывать ход своих мыслей.

---

### Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [21]:
#создадим сводную таблицу количество случаев задолжности от количества детей
data_pivot_children = data.pivot_table(index=['children'], columns='debt', aggfunc={'debt':'count'})
data_pivot_children.set_axis(['Кол-во без просрочки','Кол-во с просрочкой'], axis = 'columns', inplace=True)
data_pivot_children['%'] = (data_pivot_children['Кол-во с просрочкой'] / (data_pivot_children['Кол-во без просрочки'] + data_pivot_children['Кол-во с просрочкой']))*100
print('')
print('--------------------------------------------------')
print('Сводная таблица - Кол-во детей_Кол-во случаев просрочки')
print('--------------------------------------------------')
data_pivot_children.sort_values(by = ['%'])


--------------------------------------------------
Сводная таблица - Кол-во детей_Кол-во случаев просрочки
--------------------------------------------------


Кол-во без просрочки  Кол-во с просрочкой          %
children                                                      
0                      13028.0               1063.0   7.543822
3                        303.0                 27.0   8.181818
1                       4410.0                445.0   9.165808
2                       1858.0                194.0   9.454191
4                         37.0                  4.0   9.756098
20                        68.0                  8.0  10.526316
5                          9.0                  NaN        NaN

### Вывод

Зависимость присутствует, процент образования просроченного платежа у семей с детьми выше, чем у тех кто детей не имеет, на 1.0-2.2%. 
При этом семьи без детей гораздо чаще берут кредиты.
Семьи с 5ю детьми не допускали просрочки, но их количество слишком мало в наборе данных чтобы давать какую-то оценку.
Семьи с 20 детьми по всей видимости частные случаи, их количество мало в наборе, выводы делать невозможно.

---
## Комментарий от наставника

Вывод не противоречит полученным результатам.

---

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [22]:
#создадим сводную таблицу количество случаев задолжности от семейного полжения
data_pivot_family_status = data.pivot_table(index=['family_status'], columns='debt', aggfunc={'debt':'count'})
data_pivot_family_status .set_axis(['Кол-во без просрочки','Кол-во с просрочкой'], axis = 'columns', inplace=True)
data_pivot_family_status ['%'] = (data_pivot_family_status['Кол-во с просрочкой'] / (data_pivot_family_status['Кол-во без просрочки'] + data_pivot_family_status['Кол-во с просрочкой']))*100
print('')
print('--------------------------------------------------')
print('Сводная таблица - Семейное положение_Кол-во случаев просрочки')
print('--------------------------------------------------')
data_pivot_family_status.sort_values(by = ['%'])


--------------------------------------------------
Сводная таблица - Семейное положение_Кол-во случаев просрочки
--------------------------------------------------


Кол-во без просрочки  Кол-во с просрочкой         %
family_status                                                             
вдовец / вдова                          896                   63  6.569343
в разводе                              1110                   85  7.112971
женат / замужем                       11408                  931  7.545182
гражданский брак                       3763                  388  9.347145
не женат / не замужем                  2536                  274  9.750890

### Вывод

Семейное положение так же как и наличие детей сильно сказывается на образовании задолженности по платежам.
Выделяются люди живущие в гражданском браке и те кто не женат / не замужем, 9.3 и 9.7% соответственно среди таковых допускают образование просрочки. 
Целевая группа здесь это семейные пары, они берут гораздо больше кредитов и исправно платят, имеют высокий уровень ответственности.

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [23]:
#создадим сводную таблицу количество случаев задолжности от уровня дохода
data_pivot_income_level = data.pivot_table(index=['income_level'], columns='debt', aggfunc={'debt':'count'})
data_pivot_income_level.set_axis(['Кол-во без просрочки','Кол-во с просрочкой'], axis = 'columns', inplace=True)
data_pivot_income_level ['%'] = (data_pivot_income_level['Кол-во с просрочкой'] / (data_pivot_income_level['Кол-во без просрочки'] + data_pivot_income_level['Кол-во с просрочкой']))*100
print('')
print('--------------------------------------------------')
print('Сводная таблица - Уровень дохода_Кол-во случаев просрочки')
print('--------------------------------------------------')
data_pivot_income_level.sort_values(by = ['%'])


--------------------------------------------------
Сводная таблица - Уровень дохода_Кол-во случаев просрочки
--------------------------------------------------


Кол-во без просрочки  Кол-во с просрочкой  \
income_level                                                            
9_от 650 до 750 т. в мес.                     38                    1   
7_от 450 до 550 т. в мес.                    170                    9   
8_от 550 до 650 т. в мес.                     62                    4   
1_до 50 т. в мес.                            349                   23   
6_от 350 до 450 т. в мес.                    489                   37   
5_от 250 до 350 т. в мес.                   1815                  139   
2_от 50 до 80 т. в мес.                     1753                  151   
12_ выше 1 млн. в мес                         23                    2   
4_от 150 до 250 т. в мес.                   6449                  569   
10_от 750 до 850 т. в мес.                    11                    1   
3_от 80 до 150 т. в мес.                    8543                  804   
11_от 850 до 950 т. в мес.                     8                    1   

                                    %  
income_level                           
9_от 650 до 750 т. в мес.    2.564103  
7_от 450 до 550 т. в мес.    5.027933  
8_от 550 до 650 т. в мес.    6.060606  
1_до 50 т. в мес.            6.182796  
6_от 350 до 450 т. в мес.    7.034221  
5_от 250 до 350 т. в мес.    7.113613  
2_от 50 до 80 т. в мес.      7.930672  
12_ выше 1 млн. в мес        8.000000  
4_от 150 до 250 т. в мес.    8.107723  
10_от 750 до 850 т. в мес.   8.333333  
3_от 80 до 150 т. в мес.     8.601690  
11_от 850 до 950 т. в мес.  11.111111

### Вывод

Исходя из значений представленных в сводной таблице можно сделать вывод, что зависимость образования задолженности от уровня дохода присутствует в настоящем наборе данных.
Наиболее интересная группа, это группа '3_от 80 до 150 т. в мес' - процент просрочки - 8.6 %. Этой группе принадлежит лидерство по общему кол-ву взятых на себя денежных обязательств. Близко к группе 3 располагается группа 4, с месячным доходом находящимся между 150 и 250 тыс., процент просрочки довольно близок 8.1 %. 
В большинстве своем остальные группы имеют более низкие значения процента просрочки, но и их представительство в наборе гораздо более скромное, возможно при увеличении количества людей в этих группах, процент будет стремиться к значению 8 %.

- Как разные цели кредита влияют на его возврат в срок?

In [24]:
#создадим сводную таблицу количество случаев задолжности от целей кредита
data_pivot_purp_class = data.pivot_table(index=['purpose_class'], columns='debt', aggfunc={'debt':'count'})
data_pivot_purp_class .set_axis(['Кол-во без просрочки','Кол-во с просрочкой'], axis = 'columns', inplace=True)
data_pivot_purp_class ['%'] = (data_pivot_purp_class['Кол-во с просрочкой'] / (data_pivot_purp_class['Кол-во без просрочки'] + data_pivot_purp_class['Кол-во с просрочкой']))*100
print('')
print('--------------------------------------------------')
print('Сводная таблица - Цель кредита_Кол-во случаев просрочки')
print('--------------------------------------------------')
data_pivot_purp_class.sort_values(by = ['%'])



--------------------------------------------------
Сводная таблица - Цель кредита_Кол-во случаев просрочки
--------------------------------------------------


Кол-во без просрочки  Кол-во с просрочкой         %
purpose_class                                                     
недвижимость                  10029                  782  7.233373
свадьба                        2138                  186  8.003442
образование                    3643                  370  9.220035
автомобиль                     3903                  403  9.359034

### Вывод

Кредит на недвижимость является лидером категории, 7.2 процента от всех взятых кредитов допустили просрочки. Люди относятся к недвижимости с большей ответственностью, как правило покупая недвижимость они хотят закрыть базовую потребность человека.
В наборе данных отстутствуют сведения о суммах кредитования, думаю с ними недвижимость будет в абсолютном лидерстве.
Люди берут много кредитов на свадьбу но думаю сумма кредитования мала, от того и процент образование просрочки низок.
Автомобиль и образование - в этих цифрах внутри находится определенный процент на риск (авария, поломка, отчислили из университета и т.д.) кторый несет в себе дополнительные траты, которые не дают вносить платежи определенному проценту потребителей в срок. У недвижимости такой процент риска гораздо ниже.

---
## Комментарий от наставника

Здорово, что при анализе использован метод сводных таблиц. Он поможет в дальнейших проектах. Стоит получать сводные таблицы на данном шаге, а не в категоризации.

---

### Шаг 4. Общий вывод

Исходя из проведенного анализа, можно сделать предположение что семейная пара, без детей, берет кредит на недвижимость, будет в большем приоритете, чем любой другой набор "переменных". 
Количество детей не имеет такого веса как их принципиальное наличие да/нет.
Семейное положение женат / замужем имеет прямую влияние на возврат кредита - на это условие приходится практически полловина строк набора данных, имеет низкий процент образования долга - 7.5.
Общий вывод - на возврат кредита влияют: наличие детей, семейное положение и цель кредита.

---
## Комментарий от наставника

Выводы описаны грамотно. Приведены ответы на главные вопросы проекта. В выводах можно приводить полученные ранее значения, правильно. Также можно расписать все, что было сделано и какими методами. 

---

<font color=blue> Помарки исправлены, и теперь работа выполнена хорошо. Успехов в дальнейших проектах :)
    
---

## Комментарий от наставника

- В начале работы следует описывать данные, с которыми работаешь;
- Таблицы на печать выводи без метода print();
- Разбивай код на большее число ячеек;
- Используй альтернативный способ обработки пропусков.

Ты проделал большую работу, молодец! Работа выполнена хорошо. Подумай над альтернативным способом обработки пропусков. Жду твоей работы :)

-----

### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.